### PR TITLE
Internal code review of schedin implementation of MJARSIGNER-41 for maxTries and retryDelay

### PR DESCRIPTION
**Pull request by schedin**
This is an "internal" pull request/code review so that the code can be review by my colleague before a real pull request is created for the https://github.com/apache/maven-jarsigner-plugin project. In practice it is public, since I cloned the repo.

In this pull request I have:

1. Created a huge amount of tests
2. Used the pull request https://github.com/apache/maven-jarsigner-plugin/pull/1 by lasselindqvist as base.
3. Also integrated the suggested change by https://github.com/apache/maven-jarsigner-plugin/pull/3 by bdt-stru.

There are some refactoring need before the real pull request can be created.

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/jarsigner/AbstractJarsignerMojo.java [hunk 1]
```diff
@@ -183,6 +183,30 @@ public abstract class AbstractJarsignerMojo extends AbstractMojo {
...
+    @Parameter(property = "jarsigner.maxTries", defaultValue = "1")
+    private int maxTries;
+
+    /**
+     * Delay in seconds after a failed attempt before re-trying.
+     *
+     * @since 3.1.0
+     */
+    @Parameter(property = "jarsigner.retryDelay", defaultValue = "0")
+    private int retryDelay;
```
Comment (linustornkrantz) [inline src/main/java/org/apache/maven/plugins/jarsigner/AbstractJarsignerMojo.java]: Perhaps name this (and the Parameter property) retryDelaySeconds to clarify. 
Have you considered an exponential backoff instead? I think that a fixed delay will work fine here but it is usually a better idea to do exponential backoff.

File: src/test/java/org/apache/maven/plugins/jarsigner/PluginXmlParser.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+            if (configurationElement.hasAttribute(CONF_DEFAULT_VALUE)) {
+                String defaultValue = configurationElement.getAttribute(CONF_DEFAULT_VALUE);
+                defaultConfiguration.put(configurationParameterName, defaultValue);
+            }
+        }
+        return defaultConfiguration;
+    }
+
+    /** Searches in every &lt;mojo&gt; element for an &lt;implementation&gt; matching the class name */
```
Comment (linustornkrantz) [inline src/test/java/org/apache/maven/plugins/jarsigner/PluginXmlParser.java]: This comment contains HTML characters.

**Review by linustornkrantz — APPROVED** [review]

File: pom.xml [hunk 3]
```diff
@@ -149,6 +168,13 @@ under the License.
         </plugin>
       </plugins>
     </pluginManagement>
+    <plugins>
+      <plugin>
+        <groupId>org.jacoco</groupId>
+        <artifactId>jacoco-maven-plugin</artifactId>
+        <version>0.8.5</version>
```
Comment (schedin) [inline pom.xml]: Check version and generation

File: src/main/java/org/apache/maven/plugins/jarsigner/AbstractJarsignerMojo.java [hunk 4]
```diff
@@ -503,23 +503,23 @@ private void processArchive(final File archive) throws MojoExecutionException {
...
         }
     }
 
+    /**
+     * Executes JarSigner (execute signing/verification).
+     *
+     * @param jarSigner the JarSigner execution interface.
+     * @param request the JarSignerRequest with parameters JarSigner should use.
+     * @throws JavaToolException if java tool invocation could not be created
+     * @throws MojoExecutionException if the invocation of JarSigner went okay, but it return a non-zero exit code
```
Comment (schedin) [inline src/main/java/org/apache/maven/plugins/jarsigner/AbstractJarsignerMojo.java]: Fix language

File: src/test/java/org/apache/maven/plugins/jarsigner/TestArtifacts.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.DefaultArtifact;
+
```
Comment (schedin) [inline src/test/java/org/apache/maven/plugins/jarsigner/TestArtifacts.java:30]: Add javadoc

**Pull request merged**